src/python/pants/backend/jvm/tasks/jvm_dependency_analyzer.py

```diff
@@ -79,6 +79,21 @@ def targets_by_file(self, targets):
 
     return targets_by_file
 
+  def targets_for_class(self, target, classname):
+    """Search which targets from `target`'s transitive dependencies contain `classname`."""
+    return self._targets_by_class(tuple(target.closure())).get(classname, set())
+
+  @memoized_method
```

Comment:
This is a very expensive thing to memoize, afaict. Remember that memoizing a method treats the arguments to the method as a cache key. So you're using a very large (unsorted) tuple as a cache key here.

Instead, I think you'd want to break up the memoization so that you have per-target memoization of classnames: ie, mark a method that takes a single target `memoized_method` instead.